cli/src/commands/mod.rs

```diff
@@ -3779,17 +3783,23 @@ fn cmd_workspace_add(
         "Create initial working-copy commit in workspace {}",
         &name
     ));
-    // Check out a parent of the current workspace's working-copy commit, or the
-    // root if there is no working-copy commit in the current workspace.
-    let new_wc_commit = if let Some(old_wc_commit_id) = tx
-        .base_repo()
-        .view()
-        .get_wc_commit_id(old_workspace_command.workspace_id())
-    {
-        tx.repo().store().get_commit(old_wc_commit_id)?.parents()[0].clone()
+
+    let new_wc_commit = if let Some(specific_rev) = &args.revision {
+        old_workspace_command.resolve_single_rev(specific_rev, ui)?
     } else {
-        tx.repo().store().root_commit()
+        // Check out a parent of the current workspace's working-copy commit, or the
+        // root if there is no working-copy commit in the current workspace.
+        if let Some(old_wc_commit_id) = tx
+            .base_repo()
+            .view()
+            .get_wc_commit_id(old_workspace_command.workspace_id())
+        {
+            tx.repo().store().get_commit(old_wc_commit_id)?.parents()[0].clone()
+        } else {
+            tx.repo().store().root_commit()
+        }
```

Comment:
Couldn't this all just be replaced with `old_workspace_command.resolve_single_rev(...)` assuming that `args.revision` defaulted to `@`? Isn't the `else` branch here just a very roundabout way of saying the same thing? I'm still new to the internal API...

Reply:
Actually, no, because I don't think it would handle the `root` case correctly, and it would actually need to default to something like `@-` instead to match the current behavior, which is maybe awkward? I don't know.

Reply:
I think it would be fine with the `root` case. The reason it looks that way is to make it work if `@` doesn't exist. If you want to see what that looks like, create a throw-away clone of a repo and run `jj workspace forget` in it. (It doesn't actually have to be a throw-away repo because `jj undo` should bring you back your workspace.)